src/app/delegation_backend/src/itn_uptime_analyzer/time.go

```diff
@@ -0,0 +1,121 @@
+package itn_uptime_analyzer
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	sheets "google.golang.org/api/sheets/v4"
+)
+
+// Returns current time in UTC format
+func GetCurrentTime() time.Time {
+	currentTime := time.Now().UTC()
+	return currentTime
+}
+
+// Checks the spreadsheet for a possible value for the last execution date
+// If there is no date, or the date is between the +/- one hour range of the chosen execution interval then the function returns current time minus the execution interval
+func GetLastExecutionTime(config AppConfig, client *sheets.Service, log *logging.ZapEventLogger, sheetTitle string, currentTime time.Time, executionInterval int) time.Time {
+	readRange := fmt.Sprintf("%s!A%d:Z%d", sheetTitle, 1, 1)
+	spId := config.AnalyzerOutputGsheetId
+
+	resp, err := client.Spreadsheets.Values.Get(spId, readRange).Do()
+	if err != nil {
+		log.Fatalf("Unable to retrieve data from sheet: %v\n", err)
+	}
+
+	var lastFilledColumn int = len(resp.Values[0]) - 1
+	var lastExecutionBasedOnSheetsAsTime time.Time
+
+	readRange = fmt.Sprintf("%s!%s%d", sheetTitle, string(lastFilledColumn+65), 1)
+
+	lastTimeWindow := resp.Values[0][lastFilledColumn]
+
+	stringSplit := strings.SplitAfter(fmt.Sprint(lastTimeWindow), " - ")
+	lastExecutionBasedOnSheets := stringSplit[len(stringSplit)-1]
+
+	if !strings.HasPrefix(lastExecutionBasedOnSheets, "Node") {
+		lastExecutionBasedOnSheetsAsTime, err = time.Parse(time.RFC3339, lastExecutionBasedOnSheets)
+		if err != nil {
+			log.Fatalf("Unable to parse string to time: %v\n", err)
+		}
+	} else {
+		pastTime := currentTime.Add(-time.Duration(executionInterval) * time.Hour)
```

Comment:
This approach to calculating past time is prone to errors.

What if we have maintenance and script launches an hour later than usual and we have an unaccounted hour of identities, possibly leading to errors.

I'd suggest using the following method to handle this case:
1. Check the time of previous sheet's last column
   * This is also prone to error if someone reorders sheets manually
   * So better pick up a sheet that precedes the current by time, which would be robust if my proposal to have non-intersecting sheets would be implemented.
3. If this is the only sheet in the spreadsheet (i.e. the first run), just read all submissions to extract identities, not only from the past hour

Reply:
@georgeee the time is retrieved from the sheet, from the execution row (first row). This method of calculating the previous time is only a fallback